llama_stack/providers/inline/post_training/huggingface/recipes/finetune_multi_device.py

```diff
@@ -0,0 +1,1122 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the terms described in the LICENSE file in
+# the root directory of this source tree.
+
+import gc
+import json
+import logging
+import os
+import signal
+import sys
+from datetime import datetime, timezone
+from pathlib import Path
+from typing import Any
+
+import psutil
+import torch
+import torch.distributed as dist
+from peft import LoraConfig, LoraModel
+from torch.distributed.fsdp import FullStateDictConfig, FullyShardedDataParallel, StateDictType
+
+from llama_stack.providers.inline.post_training.common.utils import evacuate_model_from_device
+
+# Set tokenizer parallelism environment variable
+os.environ["TOKENIZERS_PARALLELISM"] = "false"
```

Comment:
These and the NCCL logging environment variables are being set for this Python process globally, right? And as soon as this module gets imported, even if the multi-device training is never used.

And, the distributed training variables in `setup_distributed_training` are effectively global as well. This opens us up to various potential concurrency issues, right? We'd need some kind of barrier to ensure only one distributed training request is ever run at a time per Llama Stack server in this setup.

Should we instead only set environment variables within the spawned processes, so that we don't pollute the environment variable space of our Llama Stack process?